zebra/rt_netlink.c

```diff
@@ -287,28 +287,24 @@ static inline int proto2zebra(int proto, int family, bool is_nexthop)
 	return proto;
 }
 
-/*
-Pending: create an efficient table_id (in a tree/hash) based lookup)
- */
 static vrf_id_t vrf_lookup_by_table(uint32_t table_id, ns_id_t ns_id)
 {
 	struct vrf *vrf;
-	struct zebra_vrf *zvrf;
+	struct zebra_vrf *zvrf, target;
 
-	RB_FOREACH (vrf, vrf_id_head, &vrfs_by_id) {
-		zvrf = vrf->info;
-		if (zvrf == NULL)
-			continue;
-		/* case vrf with netns : match the netnsid */
-		if (vrf_is_backend_netns()) {
+	if (vrf_is_backend_netns()) {
+		RB_FOREACH (vrf, vrf_id_head, &vrfs_by_id) {
+			zvrf = vrf->info;
+			if (zvrf == NULL)
+				continue;
 			if (ns_id == zvrf_id(zvrf))
 				return zvrf_id(zvrf);
-		} else {
-			/* VRF is VRF_BACKEND_VRF_LITE */
-			if (zvrf->table_id != table_id)
-				continue;
-			return zvrf_id(zvrf);
 		}
+	} else {
+		target.table_id = table_id;
+		zvrf = vrf_table_id_find(&vrfs_by_table_id, &target);
+		if (zvrf && table_id == zvrf->table_id)
```

Comment:
I don't think you need the == check right? Shouldn't your compare function of handled that?